authorizer-app-backend/src/main/java/org/radarbase/authorizer/resources/RestSourceUserResource.kt

```diff
@@ -191,23 +194,27 @@ class RestSourceUserResource(
     @Path("{id}/token")
     @NeedsPermission(Permission.Entity.MEASUREMENT, Permission.Operation.CREATE)
     fun requestToken(@PathParam("id") userId: Long): TokenDTO {
-        val user = ensureUser(userId)
-        auth.checkPermissionOnSubject(Permission.MEASUREMENT_CREATE, user.projectId, user.userId)
-        return if (user.hasValidToken()) {
-            TokenDTO(user.accessToken, user.expiresAt)
-        } else {
-            // refresh token if current token is already expired.
-            refreshToken(userId, user)
+        return lockService.runLocked("token-$userId", Duration.ofSeconds(10)) {
+            val user = ensureUser(userId)
+            auth.checkPermissionOnSubject(Permission.MEASUREMENT_CREATE, user.projectId, user.userId)
+            if (user.hasValidToken()) {
+                TokenDTO(user.accessToken, user.expiresAt)
+            } else {
+                // refresh token if current token is already expired.
+                refreshToken(userId, user)
+            }
         }
     }
 
     @POST
     @Path("{id}/token")
     @NeedsPermission(Permission.Entity.MEASUREMENT, Permission.Operation.CREATE)
     fun refreshToken(@PathParam("id") userId: Long): TokenDTO {
-        val user = ensureUser(userId)
-        auth.checkPermissionOnSubject(Permission.MEASUREMENT_CREATE, user.projectId, user.userId)
-        return refreshToken(userId, user)
+        return lockService.runLocked("token-$userId", Duration.ofSeconds(10)) {
```

Comment:
Would this be better placed inside the `refreshToken(userId: Long, user: RestSourceUser)` function. So whenever it is called, it is always locked. And then it can be placed around the refresh token part of that function (so the client does not have to wait for a lock if there is an exception to be thrown).

Reply:
The issue is that the subjects may be updated after the lock is achieved, so ensureUser always gets the latest user, including that latest tokens, etc. Indeed I could do a second user fetch.